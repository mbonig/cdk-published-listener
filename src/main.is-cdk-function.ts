import AWS from 'aws-sdk';

const sns = new AWS.SNS();
const unmarshall = AWS.DynamoDB.Converter.unmarshall;

const strip = ({ name, version }) => ({ name, version });

export const handler = async (event: any) => {
  const { dynamodb: { NewImage } } = event;

  console.log(NewImage);
  let record = unmarshall(NewImage);
  console.log(record);

  if (/aws-cdk\/core/.test(record.name) || !/aws-cdk\//.test(record.name)) {
    await sns.publish({ TopicArn: process.env.TOPIC_ARN, Message: JSON.stringify(strip(record)) }).promise();
  }
};