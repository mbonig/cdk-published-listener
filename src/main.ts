import {NodejsFunction} from '@aws-cdk/aws-lambda-nodejs';
import {Subscription, SubscriptionProtocol, Topic} from '@aws-cdk/aws-sns';
import {App, Construct, Stack, StackProps} from '@aws-cdk/core';

export class CdkPublishedListenerStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);


    const cdkPublishedTopic = new Topic(this, 'cdk-published-topic', {});

    const handler = new NodejsFunction(this, 'is-cdk-function', {
      environment: {
        TOPIC_ARN: cdkPublishedTopic.topicArn,
      },
    });
    cdkPublishedTopic.grantPublish(handler);

    let ccTopic = Topic.fromTopicArn(this, 'cc-topic', 'arn:aws:sns:us-east-1:499430655523:construct-catalog-prod-RendererTopicD9CB70E6-TTOURYQEX9K1');

    new Subscription(this, 'cc-published-subscription', {
      topic: ccTopic,
      endpoint: handler.functionArn,
      protocol: SubscriptionProtocol.LAMBDA,
    });
  }
}


const app = new App();
new CdkPublishedListenerStack(app, 'cdk-published-listener', {});
app.synth();