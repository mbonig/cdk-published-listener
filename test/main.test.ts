import '@aws-cdk/assert/jest';
import { CdkPublishedListenerStack } from '../src/main'
import { App } from '@aws-cdk/core';

test('Snapshot', () => {
  const app = new App();
  const stack = new CdkPublishedListenerStack(app, 'test');

  expect(stack).not.toHaveResource('AWS::S3::Bucket');
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});