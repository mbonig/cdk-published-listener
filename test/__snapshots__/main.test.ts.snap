// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters3da4443ffd8f5ba620239a5087486607738a56c320b0d60cad89f46725e885f4ArtifactHashD1559BF8": Object {
      "Description": "Artifact hash for asset \\"3da4443ffd8f5ba620239a5087486607738a56c320b0d60cad89f46725e885f4\\"",
      "Type": "String",
    },
    "AssetParameters3da4443ffd8f5ba620239a5087486607738a56c320b0d60cad89f46725e885f4S3Bucket43188A49": Object {
      "Description": "S3 bucket for asset \\"3da4443ffd8f5ba620239a5087486607738a56c320b0d60cad89f46725e885f4\\"",
      "Type": "String",
    },
    "AssetParameters3da4443ffd8f5ba620239a5087486607738a56c320b0d60cad89f46725e885f4S3VersionKeyFBBE4E97": Object {
      "Description": "S3 key for asset version \\"3da4443ffd8f5ba620239a5087486607738a56c320b0d60cad89f46725e885f4\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ccpublishedsubscription74B61DE8": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "iscdkfunctionC817B549",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": "arn:aws:sns:us-east-1:499430655523:construct-catalog-prod-RendererTopicD9CB70E6-TTOURYQEX9K1:5e202d75-cba1-4e5e-8e77-46f645045a93",
      },
      "Type": "AWS::SNS::Subscription",
    },
    "cdkpublishedtopicD5C468C1": Object {
      "Type": "AWS::SNS::Topic",
    },
    "iscdkfunctionC817B549": Object {
      "DependsOn": Array [
        "iscdkfunctionServiceRoleDefaultPolicyE990F27B",
        "iscdkfunctionServiceRole96BCE616",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters3da4443ffd8f5ba620239a5087486607738a56c320b0d60cad89f46725e885f4S3Bucket43188A49",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters3da4443ffd8f5ba620239a5087486607738a56c320b0d60cad89f46725e885f4S3VersionKeyFBBE4E97",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters3da4443ffd8f5ba620239a5087486607738a56c320b0d60cad89f46725e885f4S3VersionKeyFBBE4E97",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TOPIC": Object {
              "Fn::GetAtt": Array [
                "cdkpublishedtopicD5C468C1",
                "TopicName",
              ],
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "iscdkfunctionServiceRole96BCE616",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "iscdkfunctionServiceRole96BCE616": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "iscdkfunctionServiceRoleDefaultPolicyE990F27B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "cdkpublishedtopicD5C468C1",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "iscdkfunctionServiceRoleDefaultPolicyE990F27B",
        "Roles": Array [
          Object {
            "Ref": "iscdkfunctionServiceRole96BCE616",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
